File: ui-api/src/main/java/io/tipsters/uiapi/service/OddsProviderFactory.java
package io.tipsters.uiapi.service;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

import io.tipsters.uiapi.client.williamhill.OddsProvider;
import io.tipsters.uiapi.client.williamhill.OddsProviders;

import static java.util.stream.Collectors.toList;

@Service
public class OddsProviderFactory {
  private final OddsProviders oddsProviders;

  public OddsProviderFactory(OddsProviders oddsProviders) {
    this.oddsProviders = oddsProviders;
  }

  public List<OddsProvider> providersForCompetitions(Set<String> competitions) {
    List<OddsProvider> providers = oddsProviders.getProviders();
    return providers.stream()
        .filter(provider -> provider.providesCompetitions().stream().anyMatch(competitions::contains))
        .collect(toList());
  }
}

File: web-server/src/main/kotlin/io/tipsters/service/OddsProviderFactory.kt
package io.tipsters.service

import io.tipsters.common.oddsprovider.OddsProvider
import io.tipsters.config.OddsProviders
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

/**
 * Provides a mechanism to get a handle on all odds providers that support the
 * given competitions
 */
@Service
open class OddsProviderFactory @Autowired constructor(private val oddsProviders: OddsProviders) {

    fun providersForCompetitions(competitions: Set<String>): List<OddsProvider> {
        return oddsProviders.providers.filter {
            provider -> provider.providesCompetitions().any {
                competition -> competitions.contains(competition)
            }
        }
    }
}
